state_tracking.go

```diff
@@ -28,30 +30,53 @@ type StateTrackingNode struct {
 	as     string
 
 	newTracker func() stateTracker
-	groups     map[models.GroupID]*stateTrackingGroup
+
+	groupsMu sync.RWMutex
+	groups   map[models.GroupID]*stateTrackingGroup
 }
 
 func (stn *StateTrackingNode) group(g models.GroupID) (*stateTrackingGroup, error) {
+	stn.groupsMu.RLock()
 	stg := stn.groups[g]
-	if stg == nil {
-		stg = &stateTrackingGroup{}
+	stn.groupsMu.RUnlock()
```

Comment:
Just for my own knowledge, how would the node be evaluated in parallel? When I looked through the code, I didn't see anything which would result in a node's `runF` being called multiple times in parallel. Multiple different nodes, sure, but I didn't see anything for the same node.